godel_surface_detection/src/segmentation/surface_segmentation.cpp

```diff
@@ -227,71 +228,63 @@ double  SurfaceSegmentation::getSearchRadius()
 }
 
 
-bool SurfaceSegmentation::setSmoothCoef(std::vector<double> &coef)
+void SurfaceSegmentation::smoothVector(const std::vector<double> &x_in,
+                                       std::vector<double> &x_out,
+                                       const std::vector<double> &smoothing_coef)
 {
-  // smoothing filters must have an odd number of coefficients
-  if(coef.size() % 2 == 1)
+  int num_coef = smoothing_coef.size();
+  if (num_coef == 0 || num_coef % 2 != 1 || x_in.size() == 0 || x_in.size() <= num_coef)
   {
-    coef_.clear();
-    num_coef_ = coef.size();
-    double sum = 0;
-    for(const auto& c : coef)
-    {
-      coef_.push_back(c);
-      sum += c;
-    }
+    if(num_coef % 2 == 0)
+      ROS_WARN_STREAM("Filter must have an odd number of coefficients. Vector not smoothed");
 
-    // set gain to be the sum of the coefficients because we need unity gain
-    gain_ = sum;
-    return(true);
-  }
-  else
-  {
-    return(false);
+    x_out = x_in;
+    return;
   }
-}
-
 
-void SurfaceSegmentation::smoothVector(std::vector<double>&x_in, std::vector<double> &x_out)
-{
+  int gain = std::accumulate(smoothing_coef.begin(), smoothing_coef.end(), 0);
   int n = x_in.size();
-  if( n <= num_coef_)
-  {
-    x_out = x_in;
-  }
-  else
+
+  // initialize the filter using tail of x_in
+  std::vector<double> xv;
+  xv.clear();
+  for(int j = num_coef - 1; j >= 0; j--)
+    xv.push_back(x_in[n-j-1]);
+
+  // cycle through every and apply the filter
+  for(int j = 1; j < n - 1; j++)
   {
-    // initialize the filter using tail of x_in
-    std::vector<double> xv;
-    xv.clear();
-    for(int j = num_coef_ - 1; j >= 0; j--)
-      xv.push_back(x_in[n-j-1]);
-
-    // cycle through every and apply the filter
-    for(int j = 1; j < n - 1; j++)
-    {
-      // shift backwards
-      for(int k = 0; k < num_coef_ - 1; k++)
-        xv[k] = xv[k + 1];
-
-      // get next input to filter which is num_coef/2 in front of current point being smoothed
-      int idx = (j + num_coef_ / 2) % n;
-      xv[num_coef_ - 1] = x_in[idx]; // j'th point
-
-      // apply the filter
-      double sum = 0.0;
-      for(int k = 0; k < num_coef_; k++)
-        sum += xv[k] * coef_[k];
-
-      // save point
-      x_out.push_back(sum / gain_);
-    }// end for every point
-  }
+    // shift backwards
+    for(int k = 0; k < num_coef - 1; k++)
+      xv[k] = xv[k + 1];
+
+    // get next input to filter which is num_coef/2 in front of current point being smoothed
+    int idx = (j + num_coef / 2) % n;
+    xv[num_coef - 1] = x_in[idx]; // j'th point
+
+    // apply the filter
+    double sum = 0.0;
+    for(int k = 0; k < num_coef; k++)
+      sum += xv[k] * smoothing_coef[k];
+
+    // save point
+    x_out.push_back(sum / gain);
+  } // end for every point
 }
 
-
+/**
```

Comment:
I have mixed feelings about where to comment. Here I'm ascribing to the principle "put comments where they'll be read". The discussion on why the smoothing lengths are what they are and how they impact the system is relevant to the readability of the source. That said, I can put it in the header to match style/preference.